docs/docs/remark-ast.md

```diff
@@ -0,0 +1,201 @@
+
+```js
+var remark = require("remark")
+var visit = require("unist-util-visit")
+
+var tree = remark().parse("Some _emphasis_, **importance**, and `code`.")
+
+visit(tree, "text", visitor)
+
+function visitor(node) {
+  console.log(node)
+}
+```
+
+Here, it finds all text nodes and will `console.log` the nodes.
+
+Now with this, you can take the AST from your plugin and get working on adding functionality for headers:
+
+```js:title=plugins/gatsby-remark-purple-headers/index.js
+const visit = require("unist-util-visit")
+
+module.exports = ({ markdownAST }, pluginOptions) => {
+  visit(markdownAST, "heading", node => {
+    // Do stuff with heading nodes
+  })
+
+  return markdownAST
+}
+```
+
+Next, by visiting all heading nodes and passing them into a transformer function, you can manipulate the particular nodes to match your usecase.
+
+Looking again at the AST node for header:
+
+```JSON
+{
+  "type": "heading",
+  "depth": 1,
+  "children": [
+    {
+      "type": "text",
+      "value": "Hello World!",
+      "position": {
+        "start": { "line": 1, "column": 3, "offset": 2 },
+        "end": { "line": 1, "column": 15, "offset": 14 },
+        "indent": []
+      }
+    }
+  ],
+  "position": {
+    "start": { "line": 1, "column": 1, "offset": 0 },
+    "end": { "line": 1, "column": 15, "offset": 14 },
+    "indent": []
+  }
+},
+```
+
+You have context about the text as well as what depth the header is (for instance here you have a depth of 1 which would equate to an `h1` element)
+
+With the inner function of the `visit` call, you parse out all of the text and if it will map to a h1, you set the type of the node to `html` and set the node's value to be some custom HTML.
+
+```js
+const visit = require("unist-util-visit")
+const toString = require("mdast-util-to-string")
+
+module.exports = ({ markdownAST }, pluginOptions) => {
+  visit(markdownAST, "heading", node => {
+    let { depth } = node
+
+    // Skip if not an h1
+    if (depth !== 1) return
+
+    // Grab the innerText of the heading node
+    let text = toString(node)
+
+    const html = `
+        <h1 style="color: rebeccapurple">
+          ${text}
+        </h1>
+      `
+
+    node.type = "html"
+    node.children = undefined
+    node.value = html
+  })
+
+  return markdownAST
+}
+```
+
+A small library [mdast-util-to-string](https://github.com/syntax-tree/mdast-util-to-string) which is also made by Unified was used to extract the plaintext of the inner nodes, so this would remove links or other types of nodes inside the header, but given you have full access to the markdown AST, you can modify it however you wish.
+
+## Loading in changes and seeing effect
+
+After doing such, you setup some pages to be programmatically created from markdown as shown in [Part 7 of the Gatsby Tutorial](/tutorial/part-seven/). Once this is set up, you can examine that your plugin works as seen below based on the markdown you wrote earlier.
```

Comment:
```suggestion
At this point, our plugin is now ready to be used. To see the resulting functionality, it is helpful to re-visit [Part 7 of the Gatsby Tutorial](/tutorial/part-seven/) to programmatically create pages from Markdown data. Once this is set up, you can examine that your plugin works as seen below based on the markdown you wrote earlier.
```

Hmm... bit of a jump here. Perhaps we can add this as a pre-requisite or mention this earlier?

Reply:
Maybe I could add an additional section at the top of prerequisites at the top listing things that you should have under your belt such as going through the Gatsby Tutorial.

Reply:
Took the suggestion and added a prerequisite section